static/src/javascripts/projects/common/modules/experiments/ab-types.js

```diff
@@ -0,0 +1,75 @@
+// @flow
+
+type ListenerFunction = (f: () => void) => void;
+
+type VariantConfig = {
+    id: string,
+    test: () => void,
+    impression?: ListenerFunction,
+    success?: ListenerFunction,
+};
+
+type ABTestConfig = {
+    id: string,
+    start: string,
+    expiry: string,
+    author: string,
+    description: string,
+    audience: number,
+    audienceOffset: number,
+    successMeasure: string,
+    audienceCriteria: string,
+    dataLinkNames?: string,
+    idealOutcome?: string,
+    variants: Array<VariantConfig>,
+    canRun: () => boolean,
+    notInTest?: () => void,
+};
+
+export class Variant {
+    id: string;
+    test: () => void;
+    impression: ?ListenerFunction;
+    success: ?ListenerFunction;
+
+    constructor(config: VariantConfig) {
+        this.id = config.id;
+        this.test = config.test;
+        if (config.impression) this.impression = config.impression;
+        if (config.success) this.success = config.success;
+    }
+}
+
+export class ABTest {
```

Comment:
Does this need to be a `class`? If this was just an object literal with a type of `ABTestConfig`, we could use `Object.assign` to create copies of it, and we wouldn't need all this constructor logic